### PR TITLE
Fix #16995, #16994 and some other stuff ##asm

### LINKED ISSUE #16995: radare2 assemble 00 instead of the right instruction for ld hl, 0xc3a0 for gameboy
# This template is meant for bug reports, if you have a feature request, please be as descriptive as possible and delete the template

Make sure you are testing using the latest git version of radare2 before submitting any issue.

*If you would like to report a bug, please fill the template below*

### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | ubuntu 19.10 x64 
| File format of the file you reverse (mandatory)      | gameboy save file with gameboy instructions
| Architecture/bits of the file (mandatory)            | gameboy
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 24710 @ linux-x86-64 git.4.4.0-229-g5817770e2 commit: 5817770e290b3c66a3fe57850db6dda1c44c6a19 build: 2020-06-01__20:40:26

### Expected behavior

wa ld hl, 0xc3a0 should write some ocpodes obviously different than 00

### Actual behavior

wa ld hl, 0xc3a0 actually writes 00

### Steps to reproduce the behavior 
I did r2 -w myfile;
e cfg.bigendian=false; e asm.arch=gb
then wa ld hl, 0xc3a0;
then r2 still writes 00

### Additional Logs, screenshots, source-code,  configuration dump, ...

Drag and drop zip archives containing the Additional info here, don't use external services or link.

### PR DESCRIPTION
**Pull request by condret**
**<\!-- Filling this template is stupid -->**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove (haha, lol) my fix is effective or that my feature works (if possible) 
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: libr/asm/arch/gb/gbasm.c [hunk 1]
```diff
@@ -159,6 +166,38 @@ static bool gb_parse_ld2 (ut8 *buf, char *buf_asm) {
 	return false;
 }
 
+static bool gb_parse_ld3 (ut8 *buf, char *buf_asm) {
+	if (strlen (buf_asm) < 7) {
+		return false;
+	}
+	if (buf_asm[5] != ',') {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:173]: ideally it must be valid to have any unspecified number of spaces between arguments, so expecting a comma in a specific offset is probably not a good check

File: libr/asm/arch/gb/gbasm.c [hunk 2]
```diff
@@ -159,6 +166,38 @@ static bool gb_parse_ld2 (ut8 *buf, char *buf_asm) {
...
+		buf[0] = 0x21;
+		break;
+	case 0x7370:	//sp
+		buf[0] = 0x31;
+		break;
+	default:
+		return false;
+	}
+
+	const ut64 num = r_num_get (NULL, &buf_asm[6]);
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:195]: use buf_asm + 6, its easier to read and shorter

File: libr/asm/arch/gb/gbasm.c [hunk 3]
```diff
@@ -159,6 +166,38 @@ static bool gb_parse_ld2 (ut8 *buf, char *buf_asm) {
...
 
+static bool gb_parse_ld3 (ut8 *buf, char *buf_asm) {
+	if (strlen (buf_asm) < 7) {
+		return false;
+	}
+	if (buf_asm[5] != ',') {
+		return false;
+	}
+
+	const ut32 reg = (buf_asm[3] << 8) | buf_asm[4];
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:177]: use ut16

File: libr/asm/arch/gb/gbasm.c [hunk 4]
```diff
@@ -367,10 +406,16 @@ static int gbAsm(RAsm *a, RAsmOp *op, const char *buf) {
 		opbuf[0] = 0xfb;
 		break;
 	case 0x6c64: //ld
+		i = strlen (buf_asm);
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:409]: i must be size_t instead of int

File: libr/asm/arch/gb/gbasm.c [hunk 5]
```diff
@@ -515,7 +560,7 @@ static int gbAsm(RAsm *a, RAsmOp *op, const char *buf) {
 		{
 			char *p = strchr (buf_asm, (int)',');
 			if (!p) {
-				num = r_num_get (NULL, &buf_asm[3]);
+				num = r_num_get (NULL, &buf_asm[4]);
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:563]: buf_asm + 4

File: libr/asm/arch/gb/gbasm.c [hunk 6]
```diff
@@ -33,7 +33,7 @@ static bool gb_parse_cb1 (ut8 *buf, const int minlen, char *buf_asm, ut8 base) {
 		buf[0] |= (ut8)i;
 		return true;
 	}
-	if (!strncmp (buf_asm, "[hl]", 4)) {
+	if (!strncmp (&buf_asm[minlen - 1], "[hl]", 4)) {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:36]: use buf_asm + minlen - 1. Can minlen be 0? if so we have an underflow

File: libr/asm/arch/gb/gbasm.c [hunk 7]
```diff
@@ -12,7 +12,7 @@ static void str_op(char *c) {
 }
 
 static int gb_reg_idx (char r) {
-	const char *rstr = "bcdehla";
+	const char *rstr = "bcdehl a";
 	const char *ptr = strchr (rstr, r);
 	return ptr?(int)(size_t)(ptr-rstr):-1;
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:17]: add spaces around - and after :

File: libr/asm/arch/gb/gbasm.c [hunk 8]
```diff
@@ -123,17 +123,24 @@ static bool gb_parse_ld1 (ut8 *buf, const int minlen, char *buf_asm) {
 			return false;
 		}
 		buf[0] = (ut8)(0x40 + (i * 8));
-		if ((i = gb_reg_idx (buf_asm[5])) == (-1)) {
-			return false;
+		i = gb_reg_idx (buf_asm[5]);
+		if (i == (-1)) {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:127]: unnecessary ()

File: libr/asm/arch/gb/gbasm.c [hunk 9]
```diff
@@ -123,17 +123,24 @@ static bool gb_parse_ld1 (ut8 *buf, const int minlen, char *buf_asm) {
...
+			if (strncmp(&buf_asm[5], "[hl]", 4)) {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:128]: buf_asm + 5, and you can use r_str_ncasecmp() to support uppercase register namings, (and add missing space before (

File: libr/asm/arch/gb/gbasm.c (hunk 1 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:173]: yes, that is actually valid. befor ld3 is called r_str_replace_in removes multiple spaces in line 213

File: libr/asm/arch/gb/gbasm.c (hunk 6 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:36]: nope, it cannot "" is not valid instruction, also wouldn't pass through the switch-case from the caller

File: libr/asm/arch/gb/gbasm.c (hunk 7 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:17]: ok

File: libr/asm/arch/gb/gbasm.c (hunk 9 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:128]: uppercase is supported (i tested it), everything gets converted to lowercase befor

File: libr/asm/arch/gb/gbasm.c (hunk 2 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:195]: it is shorter, but harder to read

File: libr/asm/arch/gb/gbasm.c (hunk 8 above)
Comment (condret) [reply libr/asm/arch/gb/gbasm.c:127]: yes, but makes me feel safer

File: libr/asm/arch/gb/gbasm.c [hunk 10]
```diff
@@ -123,17 +125,24 @@ static bool gb_parse_ld1 (ut8 *buf, const int minlen, char *buf_asm) {
 			return false;
 		}
 		buf[0] = (ut8)(0x40 + (i * 8));
-		if ((i = gb_reg_idx (buf_asm[5])) == (-1)) {
-			return false;
+		i = gb_reg_idx (buf_asm[5]);
+		if (i == -1) {
+			if (strncmp(buf_asm + 5, "[hl]", 4)) {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:130]: missing space before (

File: libr/asm/arch/gb/gbasm.c [hunk 11]
```diff
@@ -123,17 +125,24 @@ static bool gb_parse_ld1 (ut8 *buf, const int minlen, char *buf_asm) {
...
+				return false;
+			}
+			i = 6;
 		}
 		buf[0] |= (ut8)i;
+		return true;
 	} else if (!strncmp (buf_asm + 3, "[hl],", 5)) {
```
Comment (trufae) [inline libr/asm/arch/gb/gbasm.c:137]: dont else after return

**Pull request merged**